### ISSUE TITLE
Track code stats

### ISSUE THREAD

**Issue by thorwhalen** [i2mint/isee#17]
The objective is to be able to easily check current but also historical stats (such as code quality and complexity metrics) on the state of repositories. 

Let's look for existing tools for this first. But if we needed to hack a minimal one for ourselves, here are some ideas:
* CI updates a hidden "tracks" file (json, yaml, csv,...) with the new stats on every CI run (maybe only successful ones)
* Alternative: Uses an API to tell a service to dump the stats in an actual DB. Probably Elastic, since has Kibana to view graphs etc. Easier to directly do analysis if we don't have to fetch files from each of the repos?
  
I've written a few things in [umpyre](https://pypi.org/project/umpyre/), so you can accumulate/aggregate further stuff there. 
But if possible, let's use others' packages instead of writing our own: Still, we can develop our own functional interface to these tools in `umpyre`. 

# What, and with what to track?

This should, of course, be parametrizable, because we will no doubt want to add and remove trackers. 
Implementation-wise, this points to an architecture where we can specify the what in a config file that is separate from the code that uses the config file.

Proposals for tools to use are welcome!

Here are a few ideas:
* [radon](https://pypi.org/project/radon/) computes Halstead and "Maintainability index" (and two others). 
* [coverage](https://pypi.org/project/coverage/)

**Comment by zfeng10** [comment]
Tools Ziyan’s looked into:

- `[wily](https://wily.readthedocs.io/en/latest/)` (Recommended)
    - examines the current complexity of the project or of any of the individual Python files
    - can be used in a CI/CD workflow to compare the complexity of the current files against a particular commit
    - can be used with pre-commit hooks
    - visualize change of complexity over time with table and graph
    - some commands:
        - **`wily build`:** iterate through the Git history and analyze the metrics for each file
        - **`wily report`:** see the historical trend in metrics for a given file or folder
        - **`wily graph`:** graph a set of metrics in an HTML file
        - **`wily diff`:** compares the last indexed data with the current working copy of a file
    - will detect and scan all Python code in .ipynb files automatically
    - uses `radon` to calculate metrics
   
- `[coverage](https://coverage.readthedocs.io/en/6.5.0/)` (Recommended used together with `wily`)
    - finds the lines of code that are missing tests
    - can easily run on top of `pytest`

- `[radon](https://radon.readthedocs.io/en/latest/)`
    - computes:
        - **cc**: compute Cyclomatic Complexity
        - **mi**: compute Maintainability Index
        - **hal**: compute Halstead complexity metrics
        - **raw**: compute raw metrics
            - **LOC**: the total number of lines of code
            - **LLOC**: the number of logical lines of code
            - **SLOC**: the number of source lines of code - not necessarily corresponding to the **LLOC** [[[Wikipedia]](https://radon.readthedocs.io/en/latest/commandline.html#wikipedia)](https://radon.readthedocs.io/en/latest/commandline.html#wikipedia)
            - **comments**: the number of Python comment lines (i.e. only single-line comments `#`)
            - **multi**: the number of lines representing multi-line strings
            - **blank**: the number of blank lines (or whitespace-only ones)
    - works with Jupyter Notebook
  
- `[mccabe](https://pypi.org/project/mccabe/)` - checks for McCabe code complexity (The number of independent code paths present. Want score ≤ 10
    - **1–10**: low risk, simple program;
    - **11–20**: moderate risk, more difficult program;
    - **21–50**: high risk, very difficult program;
    - **> 50**: very high risk, untestable program.
    - `flake8` wraps around
  
- `[complexity](https://github.com/thoughtbot/complexity)` - a command line tool calculates an approximation of code complexity per file in a language-agnostic way
    - not very intuitive interpretation of the values of the metrics

**Comment by thorwhalen** [comment]
Seems like wily+coverage is the way to go. @valentin-feron what do you think?

**Comment by valentin-feron** [comment]
Yes I agree. I would even say pylint+wily+coverage (code stylying + complexity tracking + test coverage)

**Comment by thorwhalen** [comment]
This need is now served by a separate, dedicated package: `umpyre` (code quality/complexity metrics tracked over time). wads's own CI template wires it in by default -- `[tool.wads.ci.metrics]` in `wads/data/pyproject_toml_tpl.toml` (`storage_branch = "code-metrics"`, per-run commits to a dedicated branch, matching the 'CI updates a hidden tracks file on every successful run' idea this issue sketched). Closing as done, by umpyre rather than by isee itself.
